compiler/coreSyn/CoreUtils.hs

```diff
@@ -145,6 +148,75 @@ coreAltsType :: [CoreAlt] -> Type
 coreAltsType (alt:_) = coreAltType alt
 coreAltsType []      = panic "corAltsType"
 
+exprUsageEnv :: CoreExpr -> UsageEnv
+exprUsageEnv = exprUsageEnv' emptyNameEnv
+
+exprUsageEnv' :: NameEnv UsageEnv -> CoreExpr -> UsageEnv
+exprUsageEnv' env (Var var) = varCallUsage env var
+exprUsageEnv' _env (Lit _) = zeroUE
+exprUsageEnv' _env (Coercion _) = zeroUE
+exprUsageEnv' env (Let bind body) =
+    exprUsageEnv' (bindUsageEnv env bind) body
+exprUsageEnv' env (Case scrut bnd _ alts) =
+    scrut_ue `addUE` altsUsageEnv env bnd alts
+  where
+    scrut_ue = varMult bnd `scaleUE` exprUsageEnv' env scrut
+exprUsageEnv' env (Cast e _) = exprUsageEnv' env e
+exprUsageEnv' env (Tick _ e) = exprUsageEnv' env e
+exprUsageEnv' env (Lam binder expr) = deleteUE (exprUsageEnv' env expr) binder
+exprUsageEnv' env (App fun (Type _)) =
+    exprUsageEnv' env fun
+exprUsageEnv' env (App fun arg) =
+    exprUsageEnv' env fun `addUE` (arg_mult `scaleUE` exprUsageEnv' env arg)
+  where
+    fun_ty = exprType fun
+    (Scaled arg_mult _, _) = splitFunTy fun_ty
+exprUsageEnv' _ _ = zeroUE
+
+bindUsageEnv :: NameEnv UsageEnv -> CoreBind -> NameEnv UsageEnv
+bindUsageEnv env (NonRec b e) = extendNameEnv env (getName b) (exprUsageEnv' env e)
+bindUsageEnv env (Rec binds) = add_ues (zip bndrs (zipWith supUE ue_anns rhs_ue))
+  where
+    add_ues ues = foldl (\e (b,ue) -> extendNameEnv e (getName b) ue) env ues
+    bndrs = map fst binds
+    rhss = map snd binds
+    ue_anns = interpUA <$> map varUsages bndrs
+    rhs_env = add_ues (zip bndrs ue_anns)
+    rhs_ue = map (exprUsageEnv' rhs_env) rhss
+
+altsUsageEnv :: NameEnv UsageEnv -> CoreBndr -> [CoreAlt] -> UsageEnv
+altsUsageEnv env b alts = supUEs $ map ((`deleteUE` b) . altUsageEnv) alts
+  where
+    altUsageEnv (DEFAULT, _, rhs) = exprUsageEnv' env rhs
+    altUsageEnv (LitAlt _, _, rhs) = exprUsageEnv' env rhs
+    altUsageEnv (DataAlt _, args, rhs) =
+      foldl deleteUE (exprUsageEnv' env rhs) args
+
+exprUsageAnnotation :: CoreExpr -> UsageAnnotation
+exprUsageAnnotation e = mkUA (exprUsageEnv e)
+
+varCallUsage :: NameEnv UsageEnv -> CoreBndr -> UsageEnv
+varCallUsage env var =
+  case lookupNameEnv env (getName var) of
+    Nothing
+      | Omega <- varMult var -> interpUA (varUsages var)
```

Comment:
This seems very strange. The `varUsages` bit should be interesting only for `let`-bound things, right? But this is guarded by a check on the `varMult`. Yet `varMult` says it is "arbitrary" for non-lambda-bound vars... but here, the arbitrary value must be `Omega`, right?